File: tweakology/Classes/UIViewController+Tweakology.m
//
//  UIViewController+Tweakology.m
//  tweakology
//
//  Created by Nikolay Ivanov on 10/15/18.
//

#import <Foundation/Foundation.h>
#import <tweakology/tweakology-Swift.h>
#import "UIViewController+Tweakology.h"
#import "SwizzlingClassProvider.h"


@implementation UIViewController (Tweakology)

+ (void)initialize {
    static dispatch_once_t once_token;
    dispatch_once(&once_token,  ^{
        SEL viewWDidLoadSelector = @selector(viewDidLoad);
        SEL viewDidLoadTweakologySelector = @selector(tweaked_viewDidLoad);
        Method originalMethod = class_getInstanceMethod(self, viewWDidLoadSelector);
        Method extendedMethod = class_getInstanceMethod(self, viewDidLoadTweakologySelector);
//        method_exchangeImplementations(originalMethod, extendedMethod);
    });
}

//- (void) tweaked_viewDidLoad {
//    [self tweaked_viewDidLoad];
//    if (@available(iOS 10.0, *)) {
//        NSDictionary *viewIndex = [self inspectLayout];
//        NSDictionary *tweaks = [TweaksStorage.sharedInstance getAllTweaks];
//        TweakologyLayoutEngine *engine = TweakologyLayoutEngine.sharedInstance;
//        [engine updateWithViewIndex:viewIndex];
//        for (id key in tweaks) {
//            NSArray *changeSeq = [tweaks objectForKey:key];
//            [engine tweakWithChangeSeq:changeSeq];
//        }
//    } else {
//        // Fallback on earlier versions
//    }
//}

@end

File: tweakology/Classes/UIViewController+Tweakology.swift
//
//  UIViewController+Tweakology.swift
//  tweakology
//
//  Created by Nikolay Ivanov on 10/14/18.
//

import Foundation

extension UIViewController {

    private static var swizzleViewDidLoadImplementation: Void {
        // make sure this isn't a subclass
//        guard self === UIViewController.self else { return }
//        DispatchQueue.once(token: "viewDidLoad") {
        let originalSelector = #selector(self.viewDidLoad)
        let swizzledSelector = #selector(self.proj_viewDidLoadSwizzled)
        let originalMethod = class_getInstanceMethod(self, originalSelector)
        let swizzledMethod = class_getInstanceMethod(self, swizzledSelector)
        method_exchangeImplementations(originalMethod!, swizzledMethod!)
//        }
    }

    @objc func proj_viewDidLoadSwizzled() {
        self.proj_viewDidLoadSwizzled()
        let viewControllerName = NSStringFromClass(type(of: self))
        print("viewDidLoad: \(viewControllerName)")
        if #available(iOS 10.0, *) {
            let viewIndex = self.inspectLayout()
            let tweaksStorage = TweaksStorage.sharedInstance
            let engine = TweakologyLayoutEngine.sharedInstance
            engine.update(viewIndex: viewIndex)
            for (_, tweakSeq) in tweaksStorage.getAllTweaks() {
                engine.tweak(changeSeq: tweakSeq)
            }
        } else {
            // Fallback on earlier versions
        }
    }

    public static func swizzleViewDidLoad() {
        _ = self.swizzleViewDidLoadImplementation
    }
}
